src/include/api/ivf_flat_index.h

```diff
@@ -0,0 +1,707 @@
+          auto x = FeatureVectorArray{std::move(s)};
+          auto y = FeatureVectorArray{std::move(t)};
+          return {std::move(x), std::move(y)};
+        }
+        default:
+          throw std::runtime_error("Unsupported attribute type");
+      }
+    }
+
+    [[nodiscard]] std::tuple<FeatureVectorArray, FeatureVectorArray>
+    query_finite_ram(
+        const QueryVectorArray& vectors,
+        size_t k_nn,
+        size_t nprobe,
+        size_t upper_bound = 0) override {
+      // @todo using index_type = size_t;
+
+      auto dtype = vectors.feature_type();
+
+      // @note We need to maintain same layout -> or swap extents
+      switch (dtype) {
+        case TILEDB_FLOAT32: {
+          auto qspan = MatrixView<float, stdx::layout_left>{
+              (float*)vectors.data(),
+              extents(vectors)[0],
+              extents(vectors)[1]};  // @todo ??
+          auto [s, t] =
+              impl_index_.query_finite_ram(qspan, k_nn, nprobe, upper_bound);
+          auto x = FeatureVectorArray{std::move(s)};
+          auto y = FeatureVectorArray{std::move(t)};
+          return {std::move(x), std::move(y)};
+        }
+        case TILEDB_UINT8: {
+          auto qspan = MatrixView<uint8_t, stdx::layout_left>{
+              (uint8_t*)vectors.data(),
+              extents(vectors)[0],
+              extents(vectors)[1]};  // @todo ??
+          auto [s, t] =
+              impl_index_.query_finite_ram(qspan, k_nn, nprobe, upper_bound);
+          auto x = FeatureVectorArray{std::move(s)};
+          auto y = FeatureVectorArray{std::move(t)};
+          return {std::move(x), std::move(y)};
+        }
+        default:
+          throw std::runtime_error("Unsupported attribute type");
+      }
+    }
+
+    // WIP
+    void update(
+        const FeatureVectorArray& vectors,
+        const std::optional<IdVector>& ids,
+        const std::optional<UpdateOptions>& options) const override {
+      //      index_.update(vectors, ids, options);
+    }
+
+    // WIP
+    void update(
+        const URI& vectors_uri,
+        const std::optional<IdVector>& ids,
+        const std::optional<UpdateOptions>& options) const override {
+      //      index_.update(vectors_uri, ids, options);
+    }
+
+    void write_index(
+        const tiledb::Context& ctx,
+        const std::string& group_uri,
+        bool overwrite) const override {
+      impl_index_.write_index(ctx, group_uri, overwrite);
+    }
+
+    // WIP
+    void remove(const IdVector& ids) const override {
+      //      index_.remove(ids);
+    }
+
+    size_t dimension() const override {
+      return ::dimension(impl_index_);
+    }
+
+    size_t num_partitions() const override {
+      return ::num_partitions(impl_index_);
+    }
+
+    // Don't think we need these
+#if 0
+    size_t ntotal() const override {
+      return ::num_vectors(impl_index_);
+    }
+
+    size_t num_vectors() const override {
+      return ::num_vectors(impl_index_);
+    }
+#endif
```

Comment:
(Minor) Remove?

Reply:
Removed `ntotal()` and `num_vectors` from both type-erased index classes.